### ISSUE TITLE
train

### ISSUE THREAD

**Issue by JC0624** [xuhongzuo/couta#4]
Hi, I have a problem during training. When I removed the backpropagation and parameter updates during training, the model was still able to train correctly and produced similar results to the paper.
![3fd67cfd49042567b66f2e96e6a8b11](https://user-images.githubusercontent.com/64187594/188372833-2e1fdadb-730e-4733-ab66-f4fa75d4ce08.png)

![1662359927067](https://user-images.githubusercontent.com/64187594/188380906-0ec0da58-c4f2-4ef5-8c54-fc7887422789.png)

**Comment by xuhongzuo** [comment]
I just tried the omi-1 entity of the ASD dataset. The detection performance is different. 


**Original:**
```
--------------------------------------------------------------------
Time: 2022-09-05 14.28.42, flag:  
Data: ASD, Algo: COUTA, Runs: 1 
Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
--------------------------------------------------------------------
 Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.28.42
|>>> epoch: 10  |   loss: 0.025457, loss_oc: 0.000454, val_loss: 0.000506
|>>> epoch: 20  |   loss: 0.013982, loss_oc: 0.000150, val_loss: 0.000168
|>>> epoch: 30  |   loss: 0.008471, loss_oc: 0.000081, val_loss: 0.000086
|>>> epoch: 40  |   loss: 0.006677, loss_oc: 0.000051, val_loss: 0.000056
ASD-omi-1,0.4468, 0.1525, 0.2130, 0.2175, 0.2086, pa, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, model, COUTA, time, 67.1 s, runs, 1/1
ASD-omi-1, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 67.1 s
f1, 0.9639, aupr, 0.9511, time, 67.1
```

after removing:
```
--------------------------------------------------------------------
Time: 2022-09-05 14.31.19, flag:  
Data: ASD, Algo: COUTA, Runs: 1 
Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
--------------------------------------------------------------------
 Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.31.19
|>>> epoch: 10  |   loss: 0.218531, loss_oc: 0.118676, val_loss: 0.118646
|>>> epoch: 20  |   loss: 0.218520, loss_oc: 0.118676, val_loss: 0.118646
|>>> epoch: 30  |   loss: 0.218528, loss_oc: 0.118676, val_loss: 0.118646
|>>> epoch: 40  |   loss: 0.218509, loss_oc: 0.118678, val_loss: 0.118646
ASD-omi-1,0.2179, 0.0921, 0.1892, 0.1045, 1.0000, pa, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, model, COUTA, time, 21.2 s, runs, 1/1
ASD-omi-1, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 21.2 s
f1, 0.7692, aupr, 0.7742, time, 21.2
```

The loss value will not change during training if you remove `loss.backward()` and `optimizer.step()`.
Which dataset did you use? Could you please double-check your code? I think there is no reason that the results will not change. There might be one exception, i.e., the used dataset is too simple, and the initialized network can work well.

**Comment by JC0624** [comment]
> 我刚刚尝试了ASD数据集的omi-1实体。检测性能不同。
> 
> **源语言：**
> 
> ```
> --------------------------------------------------------------------
> Time: 2022-09-05 14.28.42, flag:  
> Data: ASD, Algo: COUTA, Runs: 1 
> Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
> --------------------------------------------------------------------
>  Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.28.42
> |>>> epoch: 10  |   loss: 0.025457, loss_oc: 0.000454, val_loss: 0.000506
> |>>> epoch: 20  |   loss: 0.013982, loss_oc: 0.000150, val_loss: 0.000168
> |>>> epoch: 30  |   loss: 0.008471, loss_oc: 0.000081, val_loss: 0.000086
> |>>> epoch: 40  |   loss: 0.006677, loss_oc: 0.000051, val_loss: 0.000056
> ASD-omi-1,0.4468, 0.1525, 0.2130, 0.2175, 0.2086, pa, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, model, COUTA, time, 67.1 s, runs, 1/1
> ASD-omi-1, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 67.1 s
> f1, 0.9639, aupr, 0.9511, time, 67.1
> ```
> 
> 删除后：
> 
> ```
> --------------------------------------------------------------------
> Time: 2022-09-05 14.31.19, flag:  
> Data: ASD, Algo: COUTA, Runs: 1 
> Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
> --------------------------------------------------------------------
>  Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.31.19
> |>>> epoch: 10  |   loss: 0.218531, loss_oc: 0.118676, val_loss: 0.118646
> |>>> epoch: 20  |   loss: 0.218520, loss_oc: 0.118676, val_loss: 0.118646
> |>>> epoch: 30  |   loss: 0.218528, loss_oc: 0.118676, val_loss: 0.118646
> |>>> epoch: 40  |   loss: 0.218509, loss_oc: 0.118678, val_loss: 0.118646
> ASD-omi-1,0.2179, 0.0921, 0.1892, 0.1045, 1.0000, pa, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, model, COUTA, time, 21.2 s, runs, 1/1
> ASD-omi-1, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 21.2 s
> f1, 0.7692, aupr, 0.7742, time, 21.2
> ```
> 
> 如果删除 和，则在训练期间损失值不会更改。您使用了哪个数据集？你能仔细检查你的代码吗？我认为没有理由不改变结果。可能有一个例外，即使用的数据集太简单，初始化的网络可以很好地工作。`loss.backward()``optimizer.step()`

The dataset I am using is Epilepsy. I will try it on other datasets and get back to you, thanks for your reply.

**Comment by JC0624** [comment]
> > 我刚刚尝试了ASD数据集的omi-1实体。检测性能不同。
> > **源语言：**
> > ```
> > --------------------------------------------------------------------
> > Time: 2022-09-05 14.28.42, flag:  
> > Data: ASD, Algo: COUTA, Runs: 1 
> > Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
> > --------------------------------------------------------------------
> >  Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.28.42
> > |>>> epoch: 10  |   loss: 0.025457, loss_oc: 0.000454, val_loss: 0.000506
> > |>>> epoch: 20  |   loss: 0.013982, loss_oc: 0.000150, val_loss: 0.000168
> > |>>> epoch: 30  |   loss: 0.008471, loss_oc: 0.000081, val_loss: 0.000086
> > |>>> epoch: 40  |   loss: 0.006677, loss_oc: 0.000051, val_loss: 0.000056
> > ASD-omi-1,0.4468, 0.1525, 0.2130, 0.2175, 0.2086, pa, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, model, COUTA, time, 67.1 s, runs, 1/1
> > ASD-omi-1, 0.9954, 0.9511, 0.9639, 0.9304, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 67.1 s
> > f1, 0.9639, aupr, 0.9511, time, 67.1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 删除后：
> > ```
> > --------------------------------------------------------------------
> > Time: 2022-09-05 14.31.19, flag:  
> > Data: ASD, Algo: COUTA, Runs: 1 
> > Configs: {'bias': 1, 'num_epochs': 40, 'emb_dim': 32, 'hidden_dims': 32, 'rep_hidden': 32, 'lr': 0.0001, 'batch_size': 64, 'dropout': 0.0, 'alpha': 0.1, 'neg_batch_ratio': 0.2, 'es': 0} 
> > --------------------------------------------------------------------
> >  Running COUTA on ASD-omi-1 [1/1], cur_time: 2022-09-05 14.31.19
> > |>>> epoch: 10  |   loss: 0.218531, loss_oc: 0.118676, val_loss: 0.118646
> > |>>> epoch: 20  |   loss: 0.218520, loss_oc: 0.118676, val_loss: 0.118646
> > |>>> epoch: 30  |   loss: 0.218528, loss_oc: 0.118676, val_loss: 0.118646
> > |>>> epoch: 40  |   loss: 0.218509, loss_oc: 0.118678, val_loss: 0.118646
> > ASD-omi-1,0.2179, 0.0921, 0.1892, 0.1045, 1.0000, pa, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, model, COUTA, time, 21.2 s, runs, 1/1
> > ASD-omi-1, 0.9257, 0.7742, 0.7692, 0.9049, 0.6689, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 21.2 s
> > f1, 0.7692, aupr, 0.7742, time, 21.2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 如果删除 和，则在训练期间损失值不会更改。您使用了哪个数据集？你能仔细检查你的代码吗？我认为没有理由不改变结果。可能有一个例外，即使用的数据集太简单，初始化的网络可以很好地工作。`` loss.backward()``optimizer.step() ``
> 
> The dataset I am using is Epilepsy. I will try it on other datasets and get back to you, thanks for your reply.

Your idea is correct, I tried it on the WaQ dataset, and the model performance decreased overall when backpropagation was removed. thanks for your help!
